Fix preview chart in APM Latency threshold rule

benakansara opened the pull request:
Resolves the following for APM Latency threshold rule:
- https://github.com/elastic/kibana/issues/152218
- https://github.com/elastic/kibana/issues/155489

Improvements in the preview chart:
- Showing data grouped by selected group by fields
- Showing 5 groups from each bucket
- Showing last x units (minutes, hours, etc.) data
- Added loading indicator
- Showing tooltip
- Always showing legend
- Max Y is slightly higher than maximum of highest y value or threshold
- Adjusting chart to always show the threshold line
- Fixed
  - When removing some fields like transaction type, transaction name, chart showed no data
  - Hiding a particular group (by clicking on legend item), readjusted max Y value, but same was not reset to original when enabling the group back

### Before

https://github.com/elastic/kibana/assets/69037875/03b77133-a131-4758-805b-2a9287fb7ff9

### After

https://github.com/elastic/kibana/assets/69037875/5900b74e-ee26-4d1d-a3f4-a03333808075

benakansara committed:
fix preview chart in latency threshold

benakansara committed:
add tests

benakansara committed:
Merge branch 'main' into bugfix/preview-chart-apm-rules

benakansara committed:
refactoring

benakansara committed:
fix error rate preview chart

benakansara committed:
Revert "fix error rate preview chart"

This reverts commit b9b93e55af6a1385c05a15f41fdc19cd7fe2886e.

benakansara committed:
refactoring

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

benakansara committed:
Merge branch 'main' into fix/preview-chart-apm-latency-rule

benakansara committed:
preview chart improvements

benakansara committed:
fix tests

benakansara committed:
removed unused code

benakansara committed:
removing label when no data

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

benakansara committed:
Merge branch 'main' into fix/preview-chart-apm-latency-rule

sqren commented:
/oblt-deploy

sqren commented:
> Showing 3 groups from each bucket

From the video it looks like we are showing 10 groups (there are 10 legends). Am I missing something?

<img width="992" alt="image" src="https://github.com/elastic/kibana/assets/209966/d48e286f-8736-4848-a041-a875b904b2f2">

benakansara commented:
/oblt-deploy

fkanout commented on x-pack/plugins/apm/public/components/alerting/ui_components/chart_preview/chart_preview_helper.tsx:
```diff
@@ -0,0 +1,95 @@
+
+export const tooltipProps = {
+  headerFormatter: (tooltipValue: TooltipValue) =>
+    moment(tooltipValue.value).format('Y-MM-DD HH:mm:ss'),
```
@benakansara, Wouldn't it be better to use `uiSettings.get(UI_SETTINGS.DATE_FORMAT)` instead of hardcoded date format?

benakansara replied:
done

fkanout commented on x-pack/plugins/apm/server/routes/alerts/rule_types/utils/get_interval_in_seconds.ts:
```diff
@@ -0,0 +1,36 @@
+  y: 86400 * 356,
+};
+
+const NUM_BUCKETS = 5;
```
This should be imported from here https://github.com/elastic/kibana/pull/158439/files#diff-4fb21ef6e1d21c7026ceab628776c11b01770d418dcad6b594de9217bd241d1cR21

benakansara replied:
I moved this to common folder to import from both public and server.

```
export const NUM_BUCKETS_FOR_PREVIEW_CHART = 5;
```

fkanout commented on x-pack/plugins/apm/server/routes/alerts/rule_types/utils/get_interval_in_seconds.ts:
I saw that this file/getIntervalInSeconds function is used in other rules too but with no unit test, I assume it's just working 😄

benakansara replied:
Added tests 71f62fd85a503cfc9351d85b4681d33abf44918b

fkanout reviewed (changes requested):
Tested locally, and it works as expected. However, I left some comments

sqren commented on x-pack/plugins/apm/public/components/alerting/rule_types/transaction_duration_rule_type/index.tsx:
```diff
@@ -147,16 +153,21 @@ export function TransactionDurationRuleType(props: Props) {
-    />
-  ) : null;
+  const chartPreview =
+    status === FETCH_STATUS.LOADING ? (
```
You should also account for `FETCH_STATUS.NOT_INITIATED`:

```suggestion
    isPending(status) ? (
```

sqren commented on x-pack/plugins/apm/public/components/alerting/ui_components/chart_preview/chart_preview_helper.tsx:
```diff
@@ -0,0 +1,95 @@
+  }),
+};
+
+export const EmptyContainer: React.FC = ({ children }) => (
```
Does this need to be exported? I don't see it used outside this file.

benakansara replied:
It doesn't need to be exported. Updated.

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/transaction_duration/get_transaction_duration_chart_preview.ts:
```diff
@@ -51,21 +55,31 @@ export async function getTransactionDurationChartPreview({
-        ...termQuery(SERVICE_NAME, serviceName),
-        ...termQuery(TRANSACTION_TYPE, transactionType),
-        ...termQuery(TRANSACTION_NAME, transactionName),
+        ...termQuery(SERVICE_NAME, serviceName, {
+          queryEmptyString: false,
+        }),
```
Why do we want to query for empty service names?

kpatticha replied:
For the existing rules that might have an empty value. This is caused when user selected all we passed an empty string https://github.com/elastic/kibana/blob/8.6/x-pack/plugins/apm/public/components/alerting/utils/fields.tsx#L40-L44

I have fixed it in 8.8 https://github.com/elastic/kibana/blob/8.8/x-pack/plugins/apm/public/components/alerting/utils/fields.tsx#L44

sqren replied:
Ok, that makes sense. Thanks for the context! We might want to call this out explicitly - either with a code comment or better: a test that ensures that empty strings are allowed for BWC reasons.

benakansara replied:
Added tests for empty fields 7ae99b1513b0619320a3be480f83801e9074e87b

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/utils/get_all_groupby_fields.ts:
```diff
@@ -0,0 +1,29 @@
+  ruleType: string,
+  groupBy: string[] | undefined = []
+) => {
+  const predefinedGroupBy =
+    ruleType === ApmRuleType.TransactionDuration ||
+    ruleType === ApmRuleType.TransactionErrorRate
+      ? [SERVICE_NAME, SERVICE_ENVIRONMENT, TRANSACTION_TYPE]
+      : ruleType === ApmRuleType.ErrorCount
+      ? [SERVICE_NAME, SERVICE_ENVIRONMENT]
+      : [];
```
Can you use a `switch` statement here?

benakansara replied:
done

benakansara committed:
show consistent groups in preview chart

benakansara committed:
refactoring

benakansara commented:
@sqren I have made the groups consistent. Also, changed the number of groups displayed to 5 (instead of 3) and changed the chart label at the bottom to include total number of groups.

We are not showing legend in other rules, should we remove it from APM rules too?

<img width="609" alt="Screenshot 2023-06-07 at 13 19 57" src="https://github.com/elastic/kibana/assets/69037875/ad4d5868-eca1-4ae4-9f49-64cc2fc9a363">

benakansara committed:
Merge branch 'main' into fix/preview-chart-apm-latency-rule

benakansara committed:
use switch statement

benakansara committed:
fix types

sqren commented on x-pack/plugins/apm/common/rules/get_all_groupby_fields.ts:
```diff
@@ -0,0 +1,32 @@
+      break;
+  }
+
+  return union(predefinedGroupBy, groupBy);
```
nit: I first thought `groupBy` referred to to `_.groupBy` from lodash. I think a rename can make this clearer:

```suggestion
  return union(predefinedGroupByFields, customGroupByFields);
```

benakansara committed:
renaming groupBy

benakansara committed:
fix tests

sqren commented on x-pack/plugins/apm/common/rules/apm_rule_types.ts:
```diff
@@ -40,6 +40,8 @@ export enum AggregationType {
   P99 = '99th',
 }
 
+export const NUM_BUCKETS_FOR_PREVIEW_CHART = 5;
```
Ah, no. I thought this was the number of time intervals. Perhaps call it: `NUMBER_OF_TIME_SERIES_FOR_PREVIEW_CHART`

benakansara replied:
Updated.

sqren commented on x-pack/plugins/apm/public/components/alerting/ui_components/chart_preview/index.tsx:
```diff
@@ -64,36 +67,88 @@ export function ChartPreview({
+  };
+
+  const filteredSeries = useMemo(() => {
+    const sortedSeries = series.sort((a, b) => {
+      const aMax = Math.max(...a.data.map((point) => point.y as number));
+      const bMax = Math.max(...b.data.map((point) => point.y as number));
+      return bMax - aMax;
+    });
+    return sortedSeries.slice(0, NUM_BUCKETS_FOR_PREVIEW_CHART);
```
Can you use lodash `sortBy` for this?

benakansara replied:
We need to sort the series based on the Y coordinate in `series.data` in descending order. We are looking at top 5 groups with max Y values. I need to check how to apply lodash `sortBy` for this use case.

Do you have a suggestion on how to apply it? Otherwise I would like to refactor this in a separate PR.

benakansara committed:
change variable names

benakansara committed:
added tests for empty fields

benakansara committed:
added tests for lookback period

sqren commented on x-pack/plugins/apm/common/rules/get_all_groupby_fields.ts:
```diff
@@ -0,0 +1,33 @@
+  switch (ruleType) {
+    case ApmRuleType.TransactionDuration:
+    case ApmRuleType.TransactionErrorRate:
+      predefinedGroupBy = [SERVICE_NAME, SERVICE_ENVIRONMENT, TRANSACTION_TYPE];
```
Nit: I think the suffix "fields" is important to denote that this is a list of fields 

```suggestion
      predefinedGroupByFields = [SERVICE_NAME, SERVICE_ENVIRONMENT, TRANSACTION_TYPE];
```

benakansara replied:
updated

sqren commented on x-pack/plugins/apm/common/rules/get_all_groupby_fields.ts:
```diff
@@ -0,0 +1,33 @@
+
+export const getAllGroupByFields = (
+  ruleType: string,
+  ruleParamsGroupBy: string[] | undefined = []
```
nit
```suggestion
  ruleParamsGroupByFields: string[] | undefined = []
```

benakansara replied:
updated

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/transaction_duration/get_transaction_duration_chart_preview.ts:
```diff
@@ -77,6 +91,11 @@ export async function getTransactionDurationChartPreview({
 
+  const allGroupByFields = getAllGroupByFields(
+    ApmRuleType.TransactionDuration,
+    groupBy
```
```suggestion
    groupByFields
```

benakansara replied:
updated

sqren commented on x-pack/plugins/apm/server/routes/alerts/rule_types/utils/get_interval_in_seconds.ts:
```diff
@@ -0,0 +1,40 @@
+  y: 86400 * 356,
+};
+
+export const getIntervalInSeconds = (interval: string): number => {
+  const matches = interval.match(INTERVAL_STRING_RE);
+  if (matches) {
+    return (
+      parseFloat(matches[1]) *
+      INTERVAL_MULTIPLIER_FOR_LOOKBACK *
+      units[matches[2]]
+    );
+  }
+  throw new Error('Invalid interval string format.');
+};
```
This feels too low level - like something that is solved already by the platform, response ops or as a helper in AO. I'm surprised we have to do this at all - shouldn't the interval be stored as an integer, not a string?

sqren replied:
Yes, a function that converts "1m" to 60 is not apm specific. I'm surprised this is not a solved problem elsewhere (by an 3rd party lib that we already use, or by an internal elastic package)

benakansara replied:
removed `getIntervalInSeconds`

sqren commented on x-pack/test/apm_api_integration/tests/alerts/chart_preview.spec.ts:
```diff
@@ -219,8 +225,8 @@ export default function ApiTest({ getService }: FtrProviderContext) {
 
       expect(response.status).to.be(200);
       expect(response.body.latencyChartPreview[0].data[0]).to.eql({
-        x: 1627974600000,
-        y: 18485.85714285714,
+        x: 1627973700000,
+        y: 2399148.714285714,
```
Right now these values are not very useful. It's not possible for me as a reviewer to verify whether the numbers are correct. I think this is partly due to the fact that this test uses archives, instead of synthtrace. If you can change that, that would be awesome!

benakansara replied:
Updated tests to use Synthtrace.

fkanout reviewed (commented):
@benakansara, I checked out the branch and bootstrapped Kibana, but I couldn't see the chart in the rule creation flyout.
I tried to change the services, and the other rule params, but I still can't see the preview chart. Am I missing something here? 
<img width="575" alt="Screenshot 2023-06-13 at 14 53 32" src="https://github.com/elastic/kibana/assets/6838659/a95f2b82-d623-4a38-ba5f-8bb950aa10d1">

sqren commented:
> @benakansara, I checked out the branch and bootstrapped Kibana, but I couldn't see the chart in the rule creation flyout.
I tried to change the services, and the other rule params, but I still can't see the preview chart. Am I missing something here?

Did you ingest APM data?

fkanout commented:
My bad; two PRs are fixing the APM preview chart!

sqren commented:
> My bad; two PRs are fixing the APM preview chart!

Now worries. And perhaps this is a good indication that the empty state can be improve :)

benakansara commented:
@sqren We hid the preview charts in APM rules in 8.8.0. I think @fkanout was checking the PR where chart was completely hidden, not in empty state.
For empty state, the chart will show "No data to display".

fkanout reviewed (approved):
I tested it, and it works as expected, in the Alert details page and in the rule flyout. Great job, @benakansara!
<img width="1205" alt="Screenshot 2023-06-13 at 16 57 47" src="https://github.com/elastic/kibana/assets/6838659/be13068c-baaa-40ae-8266-ba5d7b284d5f">

<img width="597" alt="Screenshot 2023-06-13 at 16 52 22" src="https://github.com/elastic/kibana/assets/6838659/9cf8d49d-44f5-40e9-a2d7-d4e2c64d302e">

sqren commented:
>  I think @fkanout was checking the PR where chart was completely hidden, not in empty state. For empty state, the chart will show "No data to display".

Ahh, perfect!!

benakansara committed:
update tests

benakansara committed:
renaming

benakansara committed:
renaming

benakansara committed:
refactoring

benakansara committed:
fix types

benakansara committed:
PR feedback

benakansara committed:
refactoring

benakansara committed:
bug fix, refactoring

benakansara committed:
refactoring

benakansara committed:
move filtering of series to server

benakansara committed:
Merge branch 'main' into fix/preview-chart-apm-latency-rule

benakansara committed:
renaming

benakansara committed:
refactoring as per feedback

benakansara committed:
refactoring as per feedback

benakansara committed:
Merge branch 'main' into fix/preview-chart-apm-latency-rule

sorenlouv reviewed (approved):
all comments made. lgtm 👍

benakansara committed:
Merge branch 'main' of github.com:elastic/kibana into fix/preview-chart-apm-latency-rule
